google/cloud/bigquery_storage_v1/writer.py

```diff
@@ -364,6 +364,291 @@ def _on_rpc_done(self, future):
+                and ``proto_rows.writer_schema.proto_descriptor`` and
+                properties populated.
+            timeout:
+                How long (in seconds) to wait for the stream to be ready.
+
+        Returns:
+            A future, which can be used to process the response to the initial
+            request when it arrives.
+        """
+        with self._thread_lock:
+            return self._open(initial_request, timeout)
+
+    def _open(
+        self,
+        initial_request: gapic_types.AppendRowsRequest,
+        timeout: float = _DEFAULT_TIMEOUT,
+    ) -> AppendRowsFuture:
+        if self.is_active:
+            raise ValueError("This manager is already open.")
+
+        if self._closed:
+            raise bqstorage_exceptions.StreamClosedError(
+                "This manager has been closed and can not be re-used."
+            )
+
+        start_time = time.monotonic()
+
+        request = self._make_initial_request(initial_request)
+
+        future = AppendRowsFuture(self._stream)
+        self._queue.put(future)
+
+        self._rpc = bidi.BidiRpc(
+            self._client.append_rows,
+            initial_request=request,
+            # TODO: pass in retry and timeout. Blocked by
+            # https://github.com/googleapis/python-api-core/issues/262
+            metadata=tuple(
+                itertools.chain(
+                    self._metadata,
+                    # This header is required so that the BigQuery Storage API
+                    # knows which region to route the request to.
+                    (("x-goog-request-params", f"write_stream={self._stream_name}"),),
+                )
+            ),
+        )
+        self._rpc.add_done_callback(self._on_rpc_done)
+
+        self._consumer = bidi.BackgroundConsumer(self._rpc, self._on_response)
+        self._consumer.start()
+
+        # Make sure RPC has started before returning.
+        # Without this, consumers may get:
+        #
+        # ValueError: Can not send() on an RPC that has never been open()ed.
+        #
+        # when they try to send a request.
+        try:
+            while not self._rpc.is_active and self._consumer.is_active:
+                # Avoid 100% CPU while waiting for RPC to be ready.
+                time.sleep(_WRITE_OPEN_INTERVAL)
+
+                # TODO: Check retry.deadline instead of (per-request) timeout.
+                # Blocked by
+                # https://github.com/googleapis/python-api-core/issues/262
+                if timeout is None:
+                    continue
+                current_time = time.monotonic()
+                if current_time - start_time > timeout:
+                    break
+        except AttributeError:
+            # Handle the AttributeError which can occur if the stream is
+            # unable to be opened. In that case, self._rpc or self._consumer
+            # may be None.
+            pass
```

Comment:
This seems like it needs to be handled?  Is the scope of the try correct?  What happens if you get a failure earlier in open (e.g. if self._client.append_rows raises an exception?

Reply:
I think the try here is for a manual backoff retry to get both rpc and consumer up and running. If AttributeError is raised, it means at least one of them weren't opened successfully, and if this happens, the code will raise an exception (see line 515), so it's sort of handled right after.